### PR TITLE
Feature heapcache

### PR DESCRIPTION
**Pull request by trshaffer**
I added another allocator for the RAM cache, and can run the ATLAS benchmark on it. The config key is

    CVMFS_CACHE_RAM_MALLOC=heap

It uses the log-structured allocator @jblomer recently added. Since the new allocator needs to have id headers and know about **all** pointers to allocations, making the transactions call into the `KvStore`'s allocator was proving troublesome. This PR makes a clean separation between the scratch memory for transactions (obtained via libc's `malloc`) and the `KvStore` allocated memory. Since transactions should either succeed or fail quickly, lingering allocations that other allocators address shouldn't be a problem. For a shared cache, this separation also fits, since it might not be possible to directly allocate memory through the `KvStore`.

### CODE REVIEW

**Comment by jblomer** [comment]
Cool, many thanks!  I'm on vacation for a few days, I'll have a look in the next few days.

File: cvmfs/mountpoint.cc [hunk 1]
```diff
@@ -212,8 +212,14 @@ bool FileSystem::CreateCache() {
...
         alloc = MemoryKvStore::kMallocArena;
+      } else if (optarg == "heap") {
+        alloc = MemoryKvStore::kMallocHeap;
       }
     }
+    if (cache_bytes % 8 != 0) {
+      // Ensure 8-byte alignment for heap allocator
+      cache_bytes += 8 - (cache_bytes % 8);
+    }
     cache_bytes = max((uint64_t) 200*1024*1024, cache_bytes);
```
Comment (jblomer) [inline cvmfs/mountpoint.cc]: Not part of this PR, but it would be better to define a constant like `kDefaultRamCacheSize`

File: cvmfs/mountpoint.cc [hunk 2]
```diff
@@ -212,8 +212,14 @@ bool FileSystem::CreateCache() {
     if (options_mgr_->GetValue("CVMFS_CACHE_RAM_MALLOC", &optarg)) {
       if (optarg == "arena") {
         alloc = MemoryKvStore::kMallocArena;
+      } else if (optarg == "heap") {
+        alloc = MemoryKvStore::kMallocHeap;
       }
     }
+    if (cache_bytes % 8 != 0) {
+      // Ensure 8-byte alignment for heap allocator
+      cache_bytes += 8 - (cache_bytes % 8);
```
Comment (jblomer) [inline cvmfs/mountpoint.cc]: We have `RoundUp8` functions for 32 bit and 64 bit arguments in malloc_arena.h and malloc_heap.h.  I'd suggest to move both of these function in smalloc.h and use them from there.

**Comment by jblomer** [comment]
If the heap memory manager works well, we could remove support for the arena memory manager and system malloc.  What do you think?

File: cvmfs/malloc_heap.cc [hunk 3]
```diff
@@ -107,12 +107,7 @@ MallocHeap::MallocHeap(uint64_t capacity, CallbackPtr callback_ptr)
   assert((capacity_ % 8) == 0);
   heap_ = reinterpret_cast<unsigned char *>(sxmmap(capacity + 7));
   uintptr_t head = uintptr_t(heap_) % 8;
-  if (head > 0)
```
Comment (jblomer) [inline cvmfs/malloc_heap.cc]: The code changes here I don't quite understand.  The original intent was to ensure 8 byte alignment.  It allocates 7 bytes more and then cuts little pieces at the beginning and the end.  Actually it's probably unnecessary because `mmap()` returns something aligned on a page boundary but I didn't find a garantuee for it...

File: cvmfs/kvstore.h [hunk 4]
```diff
@@ -107,22 +111,41 @@ class MemoryKvStore :SingleCopy {
...
+    case kMallocHeap:
+      heap_ = new MallocHeap(alloc_size,
+          this->MakeCallback(&MemoryKvStore::OnBlockMove, this));
+      break;
+    default:
+      break;
     }
   }
 
   ~MemoryKvStore() {
```
Comment (jblomer) [inline cvmfs/kvstore.h]: I think allocator and deallocator became large enough to put them in the .cc file

File: cvmfs/kvstore.h [hunk 5]
```diff
@@ -107,22 +111,41 @@ class MemoryKvStore :SingleCopy {
...
     for (size_t i = 0; i < malloc_arenas_.size(); ++i) {
       delete malloc_arenas_[i];
     }
+    if (heap_) {
```
Comment (jblomer) [inline cvmfs/kvstore.h]: The shouldn't be necessary as `heap_` is initialized to `NULL`

**Comment by jblomer** [comment]
I think we should put a small header before the hash.  For a start, the header can be a single byte to store a version number.  This will simplify our life if we ever have to change the memory layout.

**Comment by jblomer** [comment]
I think it would make sense to have helper functions for the pointer arithmetic that adds/substracts the first part with the key of a memory block. Especially when we extend it with a header and we'd need to write `ptr + 1 + sizefof(shash)`.

**Comment by jblomer** [comment]
Perhaps I'm overlooking something but it looks to me that we still could use the key-value store for open transactions, too.  We could add a `Write` method to the store and set a flag in the header byte that indicates if a block is committed or in a transaction.  It would cost us more lookup calls during a transaction but we would save the memory copy on commit.

File: cvmfs/malloc_heap.cc (hunk 3 above)
Comment (trshaffer) [inline cvmfs/malloc_heap.cc]: I was running into errors with the `unmap`, which requires page-aligned pointers. Leaving the extra space mapped does waste a little at the edges.

**Comment by trshaffer** [comment]
I'd be inclined to leave glibc `malloc` in place for testing/debugging purposes, but I think the heap allocator could be the default

File: cvmfs/malloc_heap.cc (hunk 3 above)
Comment (jblomer) [inline cvmfs/malloc_heap.cc]: I see, I think we can than simply run `sxmmap` and add an assert that the resulting pointer is 8 byte aligned.

**Comment by trshaffer** [comment]
OK, I think that covers the revisions. I added a small struct for the allocation header (containing only the hash and a version number), that way the compiler can track the size and alignment. If you think helper functions would be clearer, I can add those too. `arena` alloc is gone, and `heap` is now the default. Would you prefer to make heap the exclusive allocator, @jblomer?

**Comment by jblomer** [comment]
@trshaffer I think it's all right to keep system malloc and heap.

**Comment by trshaffer** [comment]
Travis is passing, so is this ready to merge @jblomer?

**Comment by jblomer** [comment]
Very good!  I think it's ready.  I'll have one more look tomorrow and then merge.

**Comment by jblomer** [comment]
Superseded by #1685
